tests/phpunit/tests/blocks/blockHooks.php

```diff
@@ -97,4 +100,49 @@ public function test_get_hooked_blocks_matches_found() {
 			'block hooked at the last child position'
 		);
 	}
+
+	/**
+	 * @ticket 59313
+	 *
+	 * @covers serialize_blocks
+	 */
+	public function test_serialize_blocks_with_hooked_block_before() {
+		register_block_type(
+			'tests/injected-one',
+			array(
+				'block_hooks' => array(
+					'tests/hooked-at-before' => 'before',
+				),
+			)
+		);
+		register_block_type(
+			'tests/injected-two',
+			array(
+				'block_hooks' => array(
+					'tests/hooked-at-before-container' => 'before',
+				),
+			)
+		);
+
+		$content = <<<HTML
+<!-- wp:tests/hooked-at-before /-->
+<!-- wp:tests/hooked-at-before-container -->
+	<!-- wp:tests/hooked-at-before /-->
+<!-- /wp:tests/hooked-at-before-container -->
+HTML;
+		$result  = serialize_blocks(
+			parse_blocks( $content )
+		);
```

Comment:
I think we should adjust the operation that we're testing a bit: `serialize_blocks` isn't going to insert blocks unless it's given a callback that does. Doing that adds however also a level of indirection that would be good to avoid in unit testing.

With #5240 in mind, I think the following coverage would be good:
- `insert_hooked_block`: Fairly straight-forward; needs a parsed block as anchor block (that we can create directly as an array, or via `parse_blocks`). The end result then needs to be compared to the expected one; this can happen either also for parsed block arrays (although I'm not sure if PHPUnit's array comparison assertions work well for nested arrays), or for the serialized markup for both.
- `insert_hooked_blocks`: Needs the hooked blocks to be actually registered. Otherwise, similar setup to `insert_hooked_block`'s.

Reply:
Here's an example, based on your https://github.com/WordPress/gutenberg/pull/54349: 30e86630cb 😊